src/cc/lib/graph/graph.cc

```diff
@@ -156,22 +188,69 @@ void Graph::GetNodeSparseFeature(std::span<const NodeId> node_ids, std::span<con
 
     // Fill out_dimensions in case nodes don't have some features.
     std::fill(std::begin(out_dimensions), std::end(out_dimensions), 0);
-    const int64_t len = node_ids.size();
-    for (int64_t node_index = 0; node_index < len; ++node_index)
-    {
-        auto internal_id = m_node_map.find(node_ids[node_index]);
-        if (internal_id == std::end(m_node_map))
+
+    // callback function to calculate part of the full node list.
+    // sub_out_indices & sub_out_data is used to get the results for a specific part,
+    // after all parallel job finishes, these sub_out_indices & sub_out_data will be combined.
+    auto func = [this, &node_ids, &features, &out_dimensions](const int64_t &start_node_id, const int64_t &end_node_id,
+                                                              std::vector<std::vector<int64_t>> &sub_out_indices,
+                                                              std::vector<std::vector<uint8_t>> &sub_out_data) {
+        for (int64_t node_index = start_node_id; node_index < end_node_id; ++node_index)
         {
-            continue;
+            auto internal_id = m_node_map.find(node_ids[node_index]);
+            if (internal_id == std::end(m_node_map))
+            {
+                continue;
+            }
+
+            auto index = internal_id->second;
+            size_t partition_count = m_counts[index];
+            bool found = false;
+            for (size_t partition = 0; partition < partition_count && !found; ++partition, ++index)
+            {
+                found = m_partitions[m_partitions_indices[index]].GetNodeSparseFeature(
+                    m_internal_indices[index], features, node_index, out_dimensions, sub_out_indices, sub_out_data);
+            }
         }
+    };
 
-        auto index = internal_id->second;
-        size_t partition_count = m_counts[index];
-        bool found = false;
-        for (size_t partition = 0; partition < partition_count && !found; ++partition, ++index)
+    if (!m_threadPool)
+    {
+        func(0, node_ids.size(), out_indices, out_data);
+    }
+    else
+    {
+        // data_sections structure:
+        //  - job index
+        //      - feature_index
+        //          - feature_values
+        std::vector<std::vector<std::vector<int64_t>>> indice_sections;
```

Comment:
That's a lot of allocations. Can we avoid them?

Reply:
Removed these extra allocations now, currently we focus on making fixed-size dense features in parallel with thread pool.